File: aln-wrench-qt/aln/router.cpp
#include "router.h"
#include <QRandomGenerator>

Router::Router(QString address) {
    if (address.length() > 0) {
        mAddress = address;
    }
}

QStringList Router::selectServiceAddresses(QString service) {
    QStringList remoteAddresses;
    if (serviceHandlerMap.contains(service)) {
        remoteAddresses.append(mAddress);
    }
    if (serviceLoadMap.contains(service)) {
        QMap<QString, NodeLoad*> addressToLoadMap = serviceLoadMap[service];
        foreach (QString address, addressToLoadMap.keys()) {
            remoteAddresses.append(address);
        }
    }
    return remoteAddresses;
}


QString Router::selectServiceAddress(QString service) {
    if (serviceHandlerMap.contains(service))
        return mAddress;

    short minLoad = 0;
    QString remoteAddress = "";
    if (serviceLoadMap.contains(service)) {
        QMap<QString, NodeLoad*> addressToLoadMap = serviceLoadMap[service];
        foreach (QString address, addressToLoadMap.keys()) {
            NodeLoad* nodeLoad = addressToLoadMap[address];
            if (minLoad == 0 || minLoad > nodeLoad->load) {
                remoteAddress = address;
                minLoad = nodeLoad->load;
            }
        }
    }
    return remoteAddress;
}

QString Router::send(Packet* p) {
    if (p->srcAddress.length() == 0) {
        p->srcAddress = mAddress;
    }
    if (p->destAddress.length() == 0 && p->srv.length() > 0) {
        // send packet to any/all instances of the service
        QMutexLocker lock = QMutexLocker(&mMutex);
        QStringList addresses = selectServiceAddresses(p->srv);
        lock.unlock();

        if (addresses.length() > 0) {
            for (int i = 1; i < addresses.length(); i++) {
                Packet* pc = p->copy();
                pc->destAddress = addresses.at(i);
                send(pc);
            }
            p->destAddress = addresses.at(0);
        }
    }

    if (p->destAddress == mAddress) {
        PacketHandler* handler;
        QMutexLocker lock = QMutexLocker(&mMutex);
        if (serviceHandlerMap.contains(p->srv)) {
            handler = serviceHandlerMap[p->srv];
        } else if (contextHandlerMap.contains(p->ctx)) {
            handler = contextHandlerMap[p->ctx];
        } else {
            return QString("service '%1' not registered\n").arg(p->srv);
        }
        if (handler)
            handler->onPacket(p);
        lock.unlock();
        delete p;
    } else if (p->nxtAddress.length() == 0 || p->nxtAddress == mAddress) {
        if (remoteNodeMap.contains(p->destAddress)) {
            RemoteNodeInfo* rni = remoteNodeMap[p->destAddress];
            p->nxtAddress = rni->nextHop;
            rni->channel->send(p);
        }
        // TODO detect and broadcast route failure
        return "send failed; no route to " + p->destAddress;
    } else {
        return "packet is unroutable; no action taken";
    }
    return QString();
}

short Router::registerContextHandler(PacketHandler* handler) {
    QMutexLocker lock(&mMutex);
    short newCtx = QRandomGenerator::global()->generate() % ((1 << 16)-1);
    while (contextHandlerMap.contains(newCtx))
        newCtx = QRandomGenerator::global()->generate() % ((1 << 16)-1);
    contextHandlerMap.insert(newCtx, handler);
    return newCtx;
}

void Router::releaseContext(short ctx) {
    QMutexLocker lock(&mMutex);
    contextHandlerMap.remove(ctx);
}

QMap<QString, QStringList> Router::nodeServices() {
    QMap<QString, QStringList> nodeServiceMap;
    nodeServiceMap.insert(mAddress, serviceHandlerMap.keys());
    foreach (QString service, serviceLoadMap.keys()) {
        QMap<QString, NodeLoad*> loadMap = serviceLoadMap[service];
        foreach (QString address, loadMap.keys()) {
            QStringList services;
            if (nodeServiceMap.contains(address)) {
                services = nodeServiceMap.value(address);
            }
            services.append(service);
            nodeServiceMap.insert(address, services);
        }
    }
    foreach (QString address, remoteNodeMap.keys()) {
        if (!nodeServiceMap.contains(address)) {
            nodeServiceMap.insert(address, QStringList());
        }
    }
    return nodeServiceMap;
}

Packet* Router::composeNetRouteShare(QString address, short cost) {
    Packet* p = new Packet();
    p->net = Packet::NetState::ROUTE;
    p->srcAddress = mAddress;
    p->data.clear();
    QBuffer buffer(&p->data);
    buffer.open(QIODevice::Append);
    writeToBuffer(&buffer, (INT08U)address.length());
    writeToBuffer(&buffer, address);
    writeToBuffer(&buffer, (INT16U)cost);
    buffer.close();
    return p;
}

RemoteNodeInfo Router::parseNetRouteShare(Packet* p) {
    RemoteNodeInfo info;
    if (p->net != Packet::NetState::ROUTE) {
        info.err = "parseNetworkRouteSharePacket: packet.NetState != NET_ROUTE";
        return info;
    }
    if (p->data.length() == 0) {
        info.err = "parseNetworkRouteSharePacket: packet.Data is empty";
        return info;
    }

    QByteArray data = p->data;
    INT08U addrSize = data[0];

    if (data.length() != addrSize + 3) {
        info.err = QString("parseNetworkRouteSharePacket: len: %1; exp: %2").arg(data.length()).arg(addrSize + 3);
        return info;
    }

    int offset = 1;

    info.address = data.mid(offset, addrSize);
    offset += addrSize;
    info.cost = readINT16U((INT08U*)data.mid(offset, 2).data());
    info.nextHop = p->srcAddress;

    return info;
}

Packet* Router::composeNetServiceShare(QString address, QString service, short load) {
    Packet* p = new Packet();
    p->net = Packet::NetState::SERVICE;
    p->srcAddress = mAddress;
    p->data.clear();
    QBuffer buffer(&p->data, this);
    buffer.open(QIODevice::Append);
    writeToBuffer(&buffer, (INT08U)address.length());
    writeToBuffer(&buffer, address);
    writeToBuffer(&buffer, (INT08U)service.length());
    writeToBuffer(&buffer, service);
    writeToBuffer(&buffer, (INT16U)load);
    buffer.close();
    return p;
}

ServiceNodeInfo Router::parseNetServiceShare(Packet* p) {
    ServiceNodeInfo info;
    if (p->net != Packet::NetState::SERVICE) {
        info.err = "parseNetworkRouteSharePacket: packet.NetState != NET_SERVICE";
        return info;
    }
    if (p->data.length() == 0) {
        info.err = "parseNetworkRouteSharePacket: packet data is empty";
        return info;
    }

    char* data = p->data.data();
    char addrSize = data[0];
    int offset = 1;
    info.address = p->data.mid(offset, addrSize);
    offset += addrSize;

    char srvSize = data[offset];
    offset += 1;
    info.service = p->data.mid(offset, srvSize);
    offset += srvSize;

    info.load = readINT16U((INT08U*)data);
    return info;
}

Packet* Router::composeNetQuery() {
    Packet* p = new Packet();
    p->net = Packet::NetState::QUERY;
    return p;
}

void Router::removeAddress(QString address) {
    remoteNodeMap.remove(address);
    foreach(QString service, serviceLoadMap.keys())
        serviceLoadMap[service].remove(address);
}

void Router::handleNetState(Channel* channel, Packet* packet) {
    bool stateChanged = false;
    switch (packet->net) {
    case Packet::NetState::ROUTE: {
        qDebug() << QString("router '%1' recv'd ROUTE update").arg(mAddress);
        // neighbor is sharing it's routing table
        RemoteNodeInfo info = parseNetRouteShare(packet);
        if (info.err.length() > 0) {
            qDebug() << info.err;
            return;
        }

        QString msg("NET_ROUTE [%1] to:%2, via:%3, cost:%4");
        qDebug() << msg.arg(mAddress, info.address, packet->srcAddress).arg(info.cost);
        if (info.cost == 0) { // zero cost routes are removed
            if (info.address == mAddress) {
                // TODO short delay
                shareNetState();
                return;
            } else if (remoteNodeMap.contains(info.address)) {
                // remove the route

                RemoteNodeInfo* localInfo = remoteNodeMap[info.address];
                QMutexLocker lock(&mMutex);
                removeAddress(info.address);
                stateChanged = true;
                for (Channel* ch : channels) {
                    if (ch != localInfo->channel) {
                        ch->send(packet);
                    }
                }
            }
        } else { // add or update a route
            QMutexLocker lock(&mMutex);
            RemoteNodeInfo* localInfo;
            if (remoteNodeMap.contains(info.address)) {
                localInfo = remoteNodeMap[info.address];
            } else {
                localInfo = new RemoteNodeInfo();
                localInfo->address = info.address;
                remoteNodeMap.insert(info.address, localInfo);
            }

//            localInfo.lastSeen = new Date();
            if (info.cost < localInfo->cost || localInfo->cost == 0) {
                stateChanged = true;
                localInfo->channel = channel;
                localInfo->cost = info.cost;
                localInfo->nextHop = info.nextHop;
                Packet* p = composeNetRouteShare(info.address, ++info.cost);
                for (Channel* ch : channels) {
                    if (ch != channel) {
                        ch->send(p);
                    }
                }
            }
        }
    } break;

    case Packet::NetState::SERVICE: {
        qDebug() << QString("router '%1' recv'd SERVICE update").arg(mAddress);
        ServiceNodeInfo serviceInfo = parseNetServiceShare(packet);
        if (serviceInfo.err.length() > 0) {
            qDebug() << "error parsing net service: " << serviceInfo.err;
            return;
        }
        {
            QMutexLocker lock(&mMutex);
            NodeLoad* nodeLoad = new NodeLoad();
            nodeLoad->load = serviceInfo.load;
            nodeLoad->lastSeen = QDate::currentDate();
            QMap<QString, NodeLoad*> loadMap;
            if (serviceLoadMap.contains(serviceInfo.service)) {
                loadMap = serviceLoadMap[serviceInfo.service];
                if (loadMap.contains(serviceInfo.address) &&
                        loadMap[serviceInfo.address]->load == nodeLoad->load) {
                    return; // drop redunant packet to avoid propagation loops
                }
            }           
            loadMap.insert(serviceInfo.address, nodeLoad);
            serviceLoadMap.insert(serviceInfo.service, loadMap);
            stateChanged = true;
        }
        // forward the service load
        for (Channel* ch : channels) {
            if (ch != channel) {
                ch->send(packet);
            }
        }
    } break;

    case Packet::NetState::QUERY:
        qDebug() << QString("router '%1' recv'd QUERY").arg(mAddress);
        for (Packet* p : exportRouteTable())
            channel->send(p);
        for (Packet* p : exportServiceTable())
            channel->send(p);
        break;
    }

    if (stateChanged)
        emit netStateChanged();
}

void Router::onPacket(Channel* channel, Packet* packet) {
    qDebug() << "Router::onPacket from" << packet->srcAddress << ":" << QString(packet->data);
    if (packet->net != 0) {
        handleNetState(channel, packet);
    } else {
        send(packet);
    }
}

void Router::onChannelClose(Channel* channel) {
    removeChannel(channel);
}

void Router::addChannel(Channel* channel) {
    {
        QMutexLocker lock(&mMutex);
        channels.append(channel);
    }

    connect(channel, SIGNAL(packetReceived(Channel*,Packet*)), this, SLOT(onPacket(Channel*,Packet*)), Qt::QueuedConnection);
    connect(channel, SIGNAL(closing(Channel*)), this, SLOT(onChannelClose(Channel*)));

    qDebug() << QString("router '%1' sending QUERY").arg(mAddress);

    channel->send(composeNetQuery()); // immediately query the new connection
    emit channelsChanged();
}

void Router::removeChannel(Channel* ch) {
    qDebug() << "router:RemoveChannel";
    disconnect(ch, SIGNAL(packetReceived(Channel*,Packet*)), this, SLOT(onPacket(Channel*,Packet*)));
    disconnect(ch, SIGNAL(closing(Channel*)), this, SLOT(onChannelClose(Channel*)));
    {
        QMutexLocker lock(&mMutex);
        channels.remove(channels.indexOf(ch));
        // bcast the loss of routes through the channel
        foreach (QString address, remoteNodeMap.keys()) {
            qDebug() << QString("router:RemoveChannel address '%1'").arg(address);
            RemoteNodeInfo* nodeInfo = remoteNodeMap[address];
            if (nodeInfo->channel == ch) {
                removeAddress(address);
                foreach (Channel* c, channels) {
                    c->send(composeNetRouteShare(address, (short) 0));
                }
            }
        }
    }
    emit channelsChanged();
    emit netStateChanged();
}

void Router::registerService(QString service, PacketHandler* handler) {
    QMutexLocker lock(&mMutex);
    serviceHandlerMap.insert(service, handler);
}

void Router::unregisterService(QString service) {
    QMutexLocker lock(&mMutex);
    serviceHandlerMap.remove(service);
}

QList<Packet*> Router::exportRouteTable() {
    QList<Packet*> routes;
    routes.append(composeNetRouteShare(mAddress, (short) 1));
    foreach (QString address, remoteNodeMap.keys()) {
        RemoteNodeInfo* routeInfo = remoteNodeMap[address];
        routes.append(composeNetRouteShare(address, (short) (routeInfo->cost + 1)));
    }
    return routes;
}

QList<Packet*> Router::exportServiceTable() {
    QList<Packet*> services;
    foreach (QString service, serviceHandlerMap.keys()) {
        services.append(composeNetServiceShare(mAddress, service, (short) 0));
    }
    foreach (QString service, serviceLoadMap.keys()) {
        QMap<QString, NodeLoad*> loadMap = serviceLoadMap[service];
        foreach (QString address, loadMap.keys()) {
            NodeLoad* loadInfo = loadMap[address];
            services.append(composeNetServiceShare(address, service, loadInfo->load));
        }
    }
    return services;
}

void Router::shareNetState() {
    QList<Packet*> routes, services;
    {
        QMutexLocker lock(&mMutex);
        routes = exportRouteTable();
        services = exportServiceTable();
    }
    for (Channel* ch : channels) {
        for (Packet* p : routes)
            ch->send(p);
        for (Packet* p : services)
            ch->send(p);
    }
}

File: java-android/app/src/main/java/biglittleidea/aln/Router.java
package biglittleidea.aln;

import android.util.Log;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.concurrent.ThreadLocalRandom;

public class Router {

    public interface OnStateChangedListener{
        void onStateChanged();
    }

    public Router(String address) {
        this.address = address;
    }

    public class ServiceNodeInfo {
        public String service; // name of the service
        public String address; // host of the service
        public Short capacity; // remote service capacity (in arbitrary units)
        public String err; // parser error
    }

    OnStateChangedListener onStateChangedListener = null;

    // contextHandlerMap stores query response handers by contextID
    HashMap<Short, IPacketHandler> contextHandlerMap = new HashMap<Short, IPacketHandler>();

    // serviceHandlerMap maps local endpoint nodes to packet handlers
    HashMap<String, IPacketHandler> serviceHandlerMap = new HashMap<String, IPacketHandler>();

    public class RemoteNodeInfo {
        public String address; // target addres to communicate with
        public String nextHop; // next routing node for address
        public IChannel channel; // specific channel that is hosting next hop
        public short cost; // cost of using this route, generally a hop count
        public Date lastSeen; // routes should decay with a few missed updates

        public String err; // parser err msg
    }

    // map[address]RemoteNode
    HashMap<String, RemoteNodeInfo> remoteNodeMap = new HashMap<String, RemoteNodeInfo>();

    class NodeCapacity {
        short capacity;
        Date lastSeen;
    }

    // map[service][address]NodeCapacity
    HashMap<String, HashMap<String, NodeCapacity>> serviceCapacityMap = new HashMap<String, HashMap<String, NodeCapacity>>();

    // Router manages a set of channels and packet handlers
     String address = "";
    ArrayList<IChannel> channels = new ArrayList<>();

    public Set<IChannel> getChannelSet() {
        HashSet<IChannel> channelSet = new HashSet<>();
        synchronized (this) {
            for (int i = 0; i < channels.size(); i++) {
                channelSet.add(channels.get(i));
            }
        }
        return channelSet;
    }

    public String getAddress() {
        return this.address;
    }

    public void setOnStateChangedListener(OnStateChangedListener listener) {
        onStateChangedListener = listener;
    }

    private void notifyListeners() {
        if (onStateChangedListener != null) {
            onStateChangedListener.onStateChanged();
        }
    }

    public String selectService(String service) {
        if (serviceHandlerMap.containsKey(service))
            return this.address;

        // TODO select randomly with a bias based on capacity
        short maxCapacity = 0;
        String remoteAddress = "";
        if (serviceCapacityMap.containsKey(service)) {
            HashMap<String, NodeCapacity> addressToCapacityMap = serviceCapacityMap.get(service);
            for (String address : addressToCapacityMap.keySet()) {
                NodeCapacity nodeCapacity = addressToCapacityMap.get(address);
                if (maxCapacity < nodeCapacity.capacity) {
                    remoteAddress = address;
                    maxCapacity = nodeCapacity.capacity;
                }
            }
        }
        return remoteAddress;
    }

    ArrayList<String> serviceAddresses(String service) {
        ArrayList<String> addressList = new ArrayList<>();
        synchronized(this) {
            if (serviceHandlerMap.containsKey(service))
                addressList.add(this.address);

            if (serviceCapacityMap.containsKey(service)) {
                HashMap<String, NodeCapacity> addressToLoadMap = serviceCapacityMap.get(service);
                for (String address : addressToLoadMap.keySet()) {
                    addressList.add(address);
                }
            }
        }
        return addressList;
    }

    // TODO java is supposed to throw exceptions, not return error strings
    public String send(Packet p) {
        synchronized (this) {
            if (p.SourceAddress == null || p.SourceAddress.length() == 0) {
                p.SourceAddress = this.address;
            }
            if ((p.DestAddress == null || p.DestAddress.length() == 0) &&
                    (p.Service != null && p.Service.length() > 0)) {
                ArrayList<String> addresses = this.serviceAddresses(p.Service);
                if (addresses.size() > 0) {
                    for (int i = 1; i < addresses.size(); i++) {
                        Packet pc = p.copy();
                        pc.DestAddress = addresses.get(i);
                        this.send(pc);
                    }
                    p.DestAddress = addresses.get(0);
                } else {
                    return String.format("no services of type %s yet discovered on the network", p.Service);
                }
            }
        }

        if (p.DestAddress.equals(this.address)) {
            IPacketHandler handler;
            if (this.serviceHandlerMap.containsKey(p.Service)) {
                handler = this.serviceHandlerMap.get(p.Service);
                handler.onPacket(p);
            } else if (this.contextHandlerMap.containsKey(p.ContextID)) {
                handler = this.contextHandlerMap.get(p.ContextID);
                handler.onPacket(p);
            } else {
                return String.format("service '%s' not registered\n", p.Service).toString();
            }
        } else if (p.NextAddress == null || p.NextAddress.equals(this.address) || p.NextAddress.length() == 0) {
            if (this.remoteNodeMap.containsKey(p.DestAddress)) {
                RemoteNodeInfo rni = this.remoteNodeMap.get(p.DestAddress);
                p.NextAddress = rni.nextHop;
                rni.channel.send(p);
            }
        } else {
            return "packet is unroutable; no action taken";
        }
        return null;
    }

    public short registerContextHandler(IPacketHandler handler) {
        short ctx;
        synchronized (this) {
            ctx = (short) ThreadLocalRandom.current().nextInt(1, 2 << 16);
            this.contextHandlerMap.put(ctx, handler);
        }
        return ctx;
    }

    public void releaseContext(short ctx) {
        synchronized (this) {
            this.contextHandlerMap.remove(ctx);
        }
    }

    protected Packet composeNetRouteShare(String address, short cost) {
        Packet p = new Packet();
        p.Net = Packet.NetState.ROUTE;
        p.SourceAddress = this.address;
        p.Data = new byte[address.length() + 3];
        ByteBuffer buffer = ByteBuffer.allocate(Frame.BufferSize);
        buffer.put((byte) address.length());
        buffer.put(address.getBytes());
        buffer.put(Packet.writeUINT16(cost));
        buffer.rewind();
        buffer.get(p.Data);
        return p;
    }

    protected RemoteNodeInfo parseNetRouteShare(Packet packet) {
        RemoteNodeInfo info = new RemoteNodeInfo();
        if (packet.Net != Packet.NetState.ROUTE) {
            info.err = "parseNetworkRouteSharePacket: packet.NetState != NET_ROUTE";
            return info;
        }
        if (packet.Data == null || packet.Data.length == 0) {
            info.err = "parseNetworkRouteSharePacket: packet.Data is empty";
            return info;
        }

        byte[] data = packet.Data;
        byte addrSize = data[0];

        if (data.length != addrSize + 3) {
            info.err = String.format("parseNetworkRouteSharePacket: len: %d; exp: %d", data.length, addrSize + 3);
            return info;
        }

        int offset = 1;
        info.address = new String(Arrays.copyOfRange(data, offset, offset + addrSize));
        offset += addrSize;
        info.cost = Packet.readUINT16(data, offset);
        offset += 2;
        info.nextHop = packet.SourceAddress;

        return info;
    }

    protected Packet composeNetServiceShare(String address, String service, short capacity) {
        Packet p = new Packet();
        p.Net = Packet.NetState.SERVICE;
        p.SourceAddress = this.address;
        p.Data = new byte[this.address.length() + service.length() + 4];
        ByteBuffer buffer = ByteBuffer.allocate(Frame.BufferSize);
        buffer.put((byte) address.length());
        buffer.put(address.getBytes());
        buffer.put((byte) service.length());
        buffer.put(service.getBytes());
        buffer.put(Packet.writeUINT16(capacity));
        buffer.rewind();
        buffer.get(p.Data, 0, p.Data.length);
        return p;
    }

    protected ServiceNodeInfo parseNetServiceShare(Packet packet) {
        ServiceNodeInfo info = new ServiceNodeInfo();
        if (packet.Net != Packet.NetState.SERVICE) {
            info.err = "parseNetworkRouteSharePacket: packet.NetState != NET_SERVICE";
            return info;
        }
        if (packet.Data == null || packet.Data.length == 0) {
            info.err = "parseNetworkRouteSharePacket: packet.Data is empty";
            return info;
        }

        byte[] data = packet.Data;
        byte addrSize = data[0];
        int offset = 1;
        info.address = new String(Arrays.copyOfRange(data, offset, offset + addrSize));
        offset += addrSize;

        byte srvSize = data[offset];
        offset += 1;
        info.service = new String(Arrays.copyOfRange(data, offset, offset + srvSize));
        offset += srvSize;

        info.capacity = Packet.readUINT16(data, offset);
        return info;
    }

    protected Packet composeNetQuery() {
        Packet p = new Packet();
        p.Net = Packet.NetState.QUERY;
        return p;
    }

    protected void handleNetState(IChannel channel, Packet packet) {
        switch (packet.Net) {
            case Packet.NetState.ROUTE:
                // neighbor is sharing it's routing table
                RemoteNodeInfo info = parseNetRouteShare(packet);
                if (info.err != null) {
                    Log.d("ALNN", "parseNetRouteShare err: " + info.err);
                    return;
                }

                if (info.cost == 0) { // zero cost routes are removed
                    if (info.address.equals(this.address)) {
                        // fight the lie
                        try {
                            Thread.sleep(500);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        this.shareNetState();
                        return;
                    } else if (this.remoteNodeMap.containsKey(info.address)) {
                        // remove the route
                        RemoteNodeInfo localInfo = remoteNodeMap.get(info.address);
                        removeAddress(info.address);
                        for (IChannel ch : channels) {
                            if (ch != localInfo.channel) {
                                ch.send(packet);
                            }
                        }
                    }
                } else {
                    // add or update a route
                    RemoteNodeInfo localInfo;
                    synchronized (this) {
                        if (remoteNodeMap.containsKey(info.address)) {
                            localInfo = remoteNodeMap.get(info.address);
                        } else {
                            localInfo = new RemoteNodeInfo();
                            localInfo.address = info.address;
                            remoteNodeMap.put(info.address, localInfo);
                        }
                    }
                    localInfo.lastSeen = new Date();
                    if (info.cost < localInfo.cost || localInfo.cost == 0) {
                        localInfo.channel = channel;
                        localInfo.cost = info.cost;
                        localInfo.nextHop = info.nextHop;
                        Packet p = composeNetRouteShare(info.address, ++info.cost);
                        for (IChannel ch : channels) {
                            if (ch != channel) {
                                ch.send(p);
                            }
                        }
                    }
                }
                notifyListeners();
                break;

            case Packet.NetState.SERVICE:
//                Log.d("ALNN", String.format("router '%s' recv'd SERVICE update\n", address));
                ServiceNodeInfo serviceInfo = parseNetServiceShare(packet);
                if (serviceInfo.err != null) {
                    Log.d("ALNN", String.format("error parsing net service: %s\n", serviceInfo.err));
                    return;
                }
                if (serviceInfo.address == this.address) {
                    break;
                }
                synchronized (this) {
                    NodeCapacity nodeCapacity = new NodeCapacity();
                    nodeCapacity.capacity = serviceInfo.capacity;
                    nodeCapacity.lastSeen = new Date();
                    HashMap<String, NodeCapacity> capacityMap;
                    if (!serviceCapacityMap.containsKey(serviceInfo.service)) {
                        if (serviceInfo.capacity == 0)
                            return; // drop redundant packet
                        capacityMap = new HashMap<>();
                        serviceCapacityMap.put(serviceInfo.service, capacityMap);
                    } else {
                        capacityMap = serviceCapacityMap.get(serviceInfo.service);
                        if (capacityMap.containsKey(serviceInfo.address) &&
                                capacityMap.get(serviceInfo.address).capacity == nodeCapacity.capacity) {
                            return; // drop redundant packet to avoid propagation loops
                        }
                    }
                    if (serviceInfo.capacity == 0) {
                        if (capacityMap.containsKey(serviceInfo.address))
                            capacityMap.remove(serviceInfo.address);
                    } else {
                        capacityMap.put(serviceInfo.address, nodeCapacity);
                    }
                }
                // forward the service capacity
                for (IChannel ch : channels) {
                    if (ch != channel) {
                        ch.send(packet);
                    }
                }
                notifyListeners();
                break;

            case Packet.NetState.QUERY:
//                Log.d("ALNN", String.format("router '%s' recv'd QUERY\n", address));
                for (Packet p : exportRouteTable())
                    channel.send(p);
                for (Packet p : exportServiceTable())
                    channel.send(p);
                break;
        }
    }

    public void addChannel(IChannel channel) {
        synchronized (this) {
            this.channels.add(channel);
        }

        channel.receive(new IPacketHandler() {
            @Override
            public void onPacket(Packet packet) {
                if (packet.Net != 0) {
                    handleNetState(channel, packet);
                } else {
                    send(packet);
                }
            }
        }, new IChannelCloseHandler() {
            @Override
            public void onChannelClosed(IChannel ch) {
                removeChannel(ch);
            }
        });
        channel.send(composeNetQuery()); // immediately query the new connection
    }

    public void removeChannel(IChannel ch) {
        synchronized (this) {
            channels.remove(ch);
            // bcast the loss of routes through the channel
            List<String> addresses = new ArrayList<>();
            for (String address : remoteNodeMap.keySet()) {
                RemoteNodeInfo nodeInfo = remoteNodeMap.get(address);
                if (nodeInfo.channel == ch) {
                    addresses.add(address);
                }
            }
            for (String address : addresses) {
                removeAddress(address);
                for (IChannel c : channels) {
                    c.send(composeNetRouteShare(address, (short) 0));
                }
            }
        }
        notifyListeners();
    }

    protected void removeAddress(String address) {
        synchronized (this) {
            remoteNodeMap.remove(address);
            for (HashMap<String, NodeCapacity> nlm : serviceCapacityMap.values()) {
                nlm.remove(address);
            }
        }
        notifyListeners();
    }

    public void registerService(String service, IPacketHandler handler) {
        synchronized (this) {
            serviceHandlerMap.put(service, handler);
        }
        notifyListeners();

        short cap = 1;
        Packet p = composeNetServiceShare(address, service, cap);
        for (IChannel ch : channels) {
            ch.send(p);
        }
    }

    public void unregisterService(String service) {
        // TODO advertiser service remove
        synchronized (this) {
            serviceHandlerMap.remove(service);
        }
        notifyListeners();

        short cap = 0;
        Packet p = composeNetServiceShare(address, service, cap);
        for (IChannel ch : channels) {
            ch.send(p);
        }
    }

    public class ServiceListItem implements java.lang.Comparable{
        public String service;
        public short capacity;
        public ServiceListItem(String service, short capacity) {
            this.service = service;
            this.capacity = capacity;
        }

        @Override
        public int compareTo(Object o) {
            return service.compareTo(((ServiceListItem)o).service);
        }
    }

    public class NodeInfoItem {
        public String address;
        public int distance;
        public Set<ServiceListItem> services = new TreeSet<>();
        public NodeInfoItem(String address, int distance) {
            this.address = address;
            this.distance = distance;
        }
    }

    public Map<String, NodeInfoItem> availableServices() {
        TreeMap<String, NodeInfoItem> addressMap = new TreeMap<>();
        synchronized (this) {
            // initialize addressMap with remote nodes
            for (String address : remoteNodeMap.keySet()) {
                RemoteNodeInfo info = remoteNodeMap.get(address);
                NodeInfoItem item = new NodeInfoItem(address, info.cost);
                addressMap.put(address, item);
            }

            // update the address map with service
            for (String service : serviceCapacityMap.keySet()) {
                HashMap<String, NodeCapacity> addressCapacityMap = serviceCapacityMap.get(service);
                for (String address : addressCapacityMap.keySet()) {
                    if (addressCapacityMap.containsKey(address)) {
                        NodeCapacity nodenodeCapacity = addressCapacityMap.get(address);
                        addressMap.get(address).services.add(new ServiceListItem(service, nodenodeCapacity.capacity));
                    }
                }
            }

            NodeInfoItem item = new NodeInfoItem(address, 1);
            for (String service: serviceHandlerMap.keySet()) {
                item.services.add(new ServiceListItem(service, (short)1));
            }
            addressMap.put(address, item);
        }
        return addressMap;
    }

    public Packet[] exportRouteTable() {
        Packet[] routes = new Packet[remoteNodeMap.size() + 1];
        int i = 0;
        routes[i++] = composeNetRouteShare(address, (short) 1);
        for (String address : remoteNodeMap.keySet()) {
            RemoteNodeInfo routeInfo = remoteNodeMap.get(address);
            routes[i++] = composeNetRouteShare(address, (short) (routeInfo.cost + 1));
        }
        return routes;
    }

    public Packet[] exportServiceTable() {
        int sz = serviceHandlerMap.size();
        for (HashMap<String, NodeCapacity> capacityMap : serviceCapacityMap.values()) {
            sz += capacityMap.size();
        }
        Packet[] services = new Packet[sz];
        int i = 0;
        for (String service : serviceHandlerMap.keySet()) {
            services[i++] = composeNetServiceShare(this.address, service, (short) 1);
        }
        for (String service : serviceCapacityMap.keySet()) {
            HashMap<String, NodeCapacity> capacityMap = serviceCapacityMap.get(service);
            for (String address : capacityMap.keySet()) {
                NodeCapacity capcityInfo = capacityMap.get(address);
                services[i++] = composeNetServiceShare(address, service, capcityInfo.capacity);
            }
        }
        return services;
    }

    public void shareNetState() {
        Packet[] routes, services;
        synchronized (this) {
            routes = exportRouteTable();
            services = exportServiceTable();
        }
        for (IChannel ch : channels) {
            for (Packet p : routes)
                ch.send(p);
            for (Packet p : services)
                ch.send(p);
        }
    }

    public int numChannels() {
        synchronized (this) {
            return channels.size();
        }
    }
}
